deeplay/components/transformer/enc.py

```diff
@@ -0,0 +1,136 @@
+    input_features: int
+    hidden_features: Sequence[Optional[int]]
+    out_features: int
+    num_heads: int
+    blocks: LayerList[Sequential]
+
+    @property
+    def input(self):
+        """Return the input layer of the network. Equivalent to `.blocks[0]`."""
+        return self.blocks[0]
+
+    @property
+    def hidden(self):
+        """Return the hidden layers of the network. Equivalent to `.blocks[:-1]`"""
+        return self.blocks[:-1]
+
+    @property
+    def output(self):
+        """Return the last layer of the network. Equivalent to `.blocks[-1]`."""
+        return self.blocks[-1]
+
+    @property
+    def multihead(self):
+        """Return the multihead attention layer of the network. Equivalent to
+        `.blocks.multihead.layer`."""
+        return self.blocks.multihead
+
+    @property
+    def feed_forward(self):
+        """Return the feed forward layer of the network. Equivalent to
+        `.blocks.feed_forward`."""
+        return self.blocks.feed_forward
+
+    def __init__(
+        self,
+        in_features: Optional[int],
+        hidden_features: Sequence[Optional[int]],
+        out_features: int,
+        num_heads: int,
+    ):
+        super().__init__()
+
+        self.in_features = in_features
+        self.hidden_features = hidden_features
+        self.out_features = out_features
+        self.num_heads = num_heads
+
+        if out_features <= 0:
+            raise ValueError(
+                f"Number of output features must be positive, got {out_features}"
+            )
+
+        if in_features is not None and in_features <= 0:
+            raise ValueError(f"in_channels must be positive, got {in_features}")
+
+        if any(h <= 0 for h in hidden_features):
+            raise ValueError(
+                f"all hidden_channels must be positive, got {hidden_features}"
+            )
+
+        self.blocks = LayerList()
+
+        for i, (f_in, f_out) in enumerate(
+            zip([in_features, *hidden_features], [*hidden_features, out_features])
+        ):
+            self.blocks.append(
+                SequentialBlock(
+                    # First sub-block is multihead self-attention followed by
+                    # skip connection and normalization.
+                    multihead=LayerSkipNormalization(
+                        layer=MultiheadSelfAttention(
+                            f_out,
+                            num_heads,
+                            projection=Layer(nn.LazyLinear, f_out)
+                            if f_in != f_out
+                            else Layer(nn.Identity),
+                        ),
+                        skip=Add(),
+                        normalization=Layer(nn.LayerNorm, f_out),
+                    ),
+                    # Second sub-block is feed forward followed by skip connection
+                    # and normalization
+                    feed_forward=LayerSkipNormalization(
+                        layer=Sequential(
```

Comment:
May be more deeplay-like to use dl MLP here rather than torch modules in a sequential?

Reply:
Yeah, it is a good idea!

Reply:
I agree with henrik here. We can maybe think of a way to reduce visual clutter to not display identity layers as prominently

Reply:
Modified by [40bac2c](https://github.com/DeepTrackAI/deeplay/pull/69/commits/40bac2cdf4e7ddfc13914a682534ee124df3aae9)